### PR TITLE
Ben/add gradient accumulation

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/utils.py [hunk 1]
```diff
@@ -407,3 +407,45 @@ def indico_to_finetune_sequence(texts, labels=None, multi_label=True, none_value
...
+            global_step = tf.train.get_or_create_global_step()
+
+            def apply_grads():
+                with tf.control_dependencies([super(GradAccumulationOptimiser, self).apply_gradients(grads_and_accumulated_vars, *args, **kwargs)]):
+                    apply_grads_op = tf.group(zero_ops)
+                return apply_grads_op
+                
+            with tf.control_dependencies(add_gradients_ops):
+                return tf.cond(
+                    tf.equal(global_step % accum_steps, 0),
```
Comment (benleetownsend) [inline finetune/utils.py]: Applying on step 0...

File: finetune/utils.py [hunk 2]
```diff
@@ -407,3 +407,58 @@ def indico_to_finetune_sequence(texts, labels=None, multi_label=True, none_value
...
+                zero_ops.append(accum_grad.initializer)
+
+                grads_and_accumulated_vars.append((accum_grad, v))
+
+            global_step = global_step or tf.train.get_or_create_global_step()
+
+            def apply_grads():
+                with tf.control_dependencies([
+                    super(GradAccumulationOptimiser, self).apply_gradients(grads_and_accumulated_vars, global_step=global_step, *args, **kwargs)
+                ]):
```
Comment (lhz1029) [inline finetune/utils.py]: Question for my understanding: shouldn't the global step update be happening in the call to the parent apply_gradients() function? Wouldn't that mean that the global step by the time we hit the tf.cond() call the first time we run the session?

File: finetune/utils.py [hunk 3]
```diff
@@ -407,3 +407,58 @@ def indico_to_finetune_sequence(texts, labels=None, multi_label=True, none_value
...
+            for g, v in grads_and_vars:
+                accum_grad = tf.get_variable(
+                    name=g.name[:-2] + "_acc",
+                    shape=g.shape,
+                    dtype=g.dtype,
+                    initializer=tf.constant_initializer(0)
+                )
+
+                add_gradients_ops.append(accum_grad.assign_add(g))
+                zero_ops.append(accum_grad.initializer)
```
Comment (lhz1029) [inline finetune/utils.py]: What is zero_ops for again?

File: finetune/utils.py [hunk 4]
```diff
@@ -407,3 +407,58 @@ def indico_to_finetune_sequence(texts, labels=None, multi_label=True, none_value
...
+
+                grads_and_accumulated_vars.append((accum_grad, v))
+
+            global_step = global_step or tf.train.get_or_create_global_step()
+
+            def apply_grads():
+                with tf.control_dependencies([
+                    super(GradAccumulationOptimiser, self).apply_gradients(grads_and_accumulated_vars, global_step=global_step, *args, **kwargs)
+                ]):
+                    apply_grads_op = tf.group(zero_ops)
```
Comment (lhz1029) [inline finetune/utils.py]: Also, I believe I understand why we need tf.control_dependencies(add_gradients_ops) down below--i.e. we want to make sure the gradients all get added before applying them if we do meet the condition. Why do we want tf.control_dependencies before grouping the zero_ops together? Also, why do we need to group zero_ops?

File: tests/test_utils.py [hunk 5]
```diff
@@ -100,5 +101,24 @@ def test_compute_class_weights(self):
...
+        for i in range(100):
+            val_before = sess.run(loss)
+            sess.run(train_op)
+            val_after1 = sess.run(loss)
+            sess.run(train_op)
+            val_after2 = sess.run(loss)
+            self.assertEqual(val_before, val_after1)  # first step should not actually do anything
+            self.assertNotEqual(val_after1, val_after2)  # this should have applied a step and should be different.
+
+
```
Comment (lhz1029) [inline tests/test_utils.py]: Another test I'd consider adding is to make sure the gradients are indeed being added (though see my main comment about adding vs. averaging). I believe it should be easy enough by replacing opt.minimize() with opt.compute_gradients() followed by opt.apply_gradients().

File: finetune/utils.py [hunk 6]
```diff
@@ -407,3 +407,58 @@ def indico_to_finetune_sequence(texts, labels=None, multi_label=True, none_value
...
+    :param accum_steps: An int value determining how many gradients to accumulate before performing an optimizer step.
+
+    :return: A new Optimizer, with gradient accumulation.
+    """
+
+    class GradAccumulationOptimiser(optimizer_class):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+        def apply_gradients(self, grads_and_vars, global_step=None, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (benleetownsend) [reply finetune/utils.py]: Good catch!

File: finetune/utils.py (hunk 4 above)
Comment (benleetownsend) [reply finetune/utils.py]: This is a slight misuse of control_dependancies. But basically we want to apply gradients and then and only then can we reset the accumulation variables and I want a single operation that performs that sequence of events.

**Comment by benleetownsend** [comment]
BROKEN DO NOT MERGE
